pkg/instancetype/controller/vm/controller_test.go

```diff
@@ -0,0 +1,1052 @@
+	}
+
+	sanitySync := func(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) {
+		stores := []cache.Store{
+			instancetypeInformerStore,
+			clusterInstancetypeInformerStore,
+			preferenceInformerStore,
+			clusterPreferenceInformerStore,
+			controllerrevisionInformerStore,
+		}
+		listOfObjects := [][]interface{}{}
+
+		for _, store := range stores {
+			listOfObjects = append(listOfObjects, deepCopyList(store.List()))
+		}
+
+		_, err := instancetypeController.Sync(vm, vmi)
+		Expect(err).ToNot(HaveOccurred())
+
+		for i, objects := range listOfObjects {
+			ExpectWithOffset(1, stores[i].List()).To(ConsistOf(objects...))
+		}
+	}
+
+	Context("instancetype", func() {
+		var clusterInstancetypeObj *v1beta1.VirtualMachineClusterInstancetype
+
+		BeforeEach(func() {
+			clusterInstancetypeObj = &v1beta1.VirtualMachineClusterInstancetype{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:       "clusterInstancetype",
+					UID:        resourceUID,
+					Generation: resourceGeneration,
+				},
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: v1beta1.SchemeGroupVersion.String(),
+					Kind:       "VirtualMachineClusterInstancetype",
+				},
+				Spec: v1beta1.VirtualMachineInstancetypeSpec{
+					CPU: v1beta1.CPUInstancetype{
+						Guest: uint32(2),
+					},
+					Memory: v1beta1.MemoryInstancetype{
+						Guest: resource.MustParse("128M"),
+					},
+				},
+			}
+			_, err := virtClient.VirtualMachineClusterInstancetype().Create(
+				context.Background(), clusterInstancetypeObj, metav1.CreateOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			err = clusterInstancetypeInformerStore.Add(clusterInstancetypeObj)
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("should store VirtualMachineInstancetype as ControllerRevision on Sync", func() {
+			vm.Spec.Instancetype = &virtv1.InstancetypeMatcher{
+				Name: instancetypeObj.Name,
+				Kind: instancetypeapi.SingularResourceName,
+			}
+
+			var err error
+			vm, err = virtClient.VirtualMachine(vm.Namespace).Create(context.TODO(), vm, metav1.CreateOptions{})
+			Expect(err).To(Succeed())
+
+			expectedRevisionName := instancetype.GetRevisionName(
+				vm.Name, instancetypeObj.Name, instancetypeObj.GroupVersionKind().Version, instancetypeObj.UID, instancetypeObj.Generation)
+			expectedRevision, err := instancetype.CreateControllerRevision(vm, instancetypeObj)
+			Expect(err).ToNot(HaveOccurred())
+
+			sanitySync(vm, vmi)
+
+			vm, err = virtClient.VirtualMachine(vm.Namespace).Get(context.TODO(), vm.Name, metav1.GetOptions{})
+			Expect(err).To(Succeed())
+			Expect(vm.Spec.Instancetype.RevisionName).To(Equal(expectedRevision.Name))
+
+			revision, err := virtClient.AppsV1().ControllerRevisions(vm.Namespace).Get(
+				context.Background(), expectedRevisionName, metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			revisionInstancetype, ok := revision.Data.Object.(*v1beta1.VirtualMachineInstancetype)
+			Expect(ok).To(BeTrue(), "Expected Instancetype in ControllerRevision")
```

Comment:
Should we compare contents as well?

Reply:
Or is it covered by the table below?

Reply:
Contents of the ControllerRevision? We compare that below on L278.